src/main/java/org/traccar/protocol/GatorProtocolEncoder.java

```diff
@@ -0,0 +1,192 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.traccar.protocol;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.traccar.BaseProtocolEncoder;
+import org.traccar.Main;
+import org.traccar.Protocol;
+import org.traccar.helper.Checksum;
+import org.traccar.model.Command;
+
+public class GatorProtocolEncoder extends BaseProtocolEncoder {
+
+    public GatorProtocolEncoder(Protocol protocol) {
+        super(protocol);
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(Main.class);
+
+    public static final int GET_POSITION = 0x30;
+    public static final int SET_ENGINE_START = 0x38;
+    public static final int SET_ENGINE_STOP = 0x39;
+
+
+    private ByteBuf engineExecute(String pseudoAddress, int engineState) {
+        ByteBuf buf = Unpooled.buffer(256);
+
+        // Add Header - 24 24
+        buf.writeByte(0x24);
+        buf.writeByte(0x24);
+
+        // Add Main Order - 0x38 for Engine Start, 0x39 for Engine Stop
+        buf.writeByte(engineState);
+
+        // Add Packet Length - 00 06
+        buf.writeByte(0x00);
+        buf.writeByte(0x06);
+
+        // Add Device ID
+        int[] ipAddress = numToIp(pseudoAddress);
+
+        buf.writeByte(ipAddress[0]);
+        buf.writeByte(ipAddress[1]);
+        buf.writeByte(ipAddress[2]);
+        buf.writeByte(ipAddress[3]);
+
+        // Add Checksum
+        buf.writeByte(Checksum.xor(buf.nioBuffer(2, buf.writerIndex())));
+
+        // End of Packet -> 0D
+        buf.writeByte(0x0D);
+
+        return buf;
+    }
+
+    private ByteBuf getFromDevice(String pseudoAddress, int type) {
+        ByteBuf buf = Unpooled.buffer(256);
+
+        // Add Header - 24 24
+        buf.writeByte(0x24);
+        buf.writeByte(0x24);
+
+        // Add Type
+        buf.writeByte(type);
+
+        // Add Packet Length - 00 06
+        buf.writeByte(0x00);
+        buf.writeByte(0x06);
+
+        // Add Device ID
+        int[] ipAddress = numToIp(pseudoAddress);
+
+        buf.writeByte(ipAddress[0]);
+        buf.writeByte(ipAddress[1]);
+        buf.writeByte(ipAddress[2]);
+        buf.writeByte(ipAddress[3]);
+
+        // Add Checksum
+        buf.writeByte(Checksum.xor(buf.nioBuffer(2, buf.writerIndex())));
+
+        // End of Packet -> 0D
+        buf.writeByte(0x0D);
+
+        return buf;
+    }
+
+    public int[] numToIp(String sim) {
```

Comment:
This function doesn't look right. First of all, the naming of the function and variables is not per our standards. Also, it looks too complicated. It just doesn't make sense. What are you trying to do here?

Reply:
There has to be a better way. This is what it looks like in the decoder:

https://github.com/traccar/traccar/blob/master/src/main/java/org/traccar/protocol/GatorProtocolDecoder.java#L51-L60

Reply:
@tananaev 

I simplified the encoding function.
Tested it with my actual devices.

https://github.com/abish7643/traccar/blob/Gator_Protocol_Improved/src/main/java/org/traccar/protocol/GatorProtocolEncoder.java#L82-L98